src/core/ucc_context.c

```diff
@@ -226,6 +275,14 @@ ucc_status_t ucc_context_create(ucc_lib_h lib,
     ctx->lib = lib;
     ucc_copy_context_params(&ctx->params, params);
     ucc_copy_context_params(&b_params.params, params);
+    b_params.context = ctx;
+    status = ucc_create_tl_contexts(ctx, config);
```

Comment:
I’m wondering in this approach, where you initialize multiple CLs but you end up using only one CL, this will result in creating contexts for all TLs? Would wrapping TL under CL help overcome this? 
Did you explore that approach? Any obvious disadvantages?

Reply:
yes, i've considered that approach: see closed/rejected PR #53 - we went over that approach with Sergey. Shortly, the downsides are the complexity of implementation and maintenance (need to keep hash-table based storage for tl lib and context objects in order to organize TL ctx sharing between CLs) and somewhat more complex cleanup flow. These complexities could be fine if we would get some real benefit from that approach, but looks like we actually don't. 

Regarding you example: we first filter CL lib objects based on user input, only after that we build the list of required TLs. So, the only corner case is : when some CL is supposed to be used but it fails creating its context AND it requested some TL which is not used by other CLs. In that case there is some TL context which is created w/o the need (so we will need to destroy it right away - this is TODO). However, this corner case still exist even in that alternative approach: consider some CL (say CL_HW - that implements a HW collectives) that needs a specific TL (say  sharp) and this CL allocates TL context internally when cl_context is being initialized. If the cl_context create fails for some reason then that CL has to cleanup allocated TL anyways. So, it is all the same but happening in different place.

Lastly, third argument: there is almost no reason for CL context to fail during init except for TL context init failure (this is where resources are allocated) unless some critical error (like OOM). 

So, i believe the approach taken in this PR is cleaner and has basically same corner cases as alternative approach.